mlflow/_spark_autologging.py

```diff
@@ -0,0 +1,230 @@
+    repl_id = SparkContext.getOrCreate().getLocalProperty("spark.databricks.replId")
+    if repl_id:
+        return repl_id
+    main_file = sys.argv[0] if len(sys.argv) > 0 else "<console>"
+    return "PythonSubscriber[{filename}][{id}]".format(filename=main_file, id=uuid.uuid4().hex)
+
+
+class PythonSubscriber(object):
+    """
+    Subscriber, intended to be instantiated once per Python process, that logs Spark table
+    information propagated from Java to the current MLflow run, starting a run if necessary.
+    class implements a Java interface (org.mlflow.spark.autologging.MlflowAutologEventSubscriber,
+    defined in the mlflow-spark package) that's called-into by autologging logic in the JVM in order
+    to propagate Spark datasource read events to Python.
+    """
+    def __init__(self):
+        self._repl_id = _get_repl_id()
+
+    def toString(self):
+        # For debugging
+        return "PythonSubscriber<replId=%s>" % self.replId()
+
+    def ping(self):
+        return None
+
+    def notify(self, path, version, data_format):
+        try:
+            self._notify(path, version, data_format)
+        except Exception as e:  # pylint: disable=broad-except
+            _logger.error("Unexpected exception %s while attempting to log Spark datasource "
+                          "info. Exception:\n", e)
+
+    def _notify(self, path, version, data_format):
+        """
+        Method called by Scala SparkListener to propagate datasource read events to the current
+        Python process
+        """
+        # If there's an active run, simply set the tag on it
+        # Note that there's a TOCTOU race condition here - active_run() here can actually throw
+        # if the main thread happens to end the run & pop from the active run stack after we check
+        # the stack size but before we peek
+        active_run = mlflow.active_run()
+        if active_run:
+            _set_run_tag_async(active_run.info.run_id, path, version, data_format)
+        else:
+            add_table_info_to_context_provider(path, version, data_format)
+
+    def register(self):
+        event_publisher = _get_jvm_event_publisher()
+        event_publisher.register(self)
+
+    def replId(self):
+        return self._repl_id
+
+    class Java:
+        implements = ["{}.MlflowAutologEventSubscriber".format(_JAVA_PACKAGE)]
+
+
+class SparkAutologgingContext(RunContextProvider):
+    """
+    Context provider used when there's no active run. Accumulates datasource read information,
+    then logs that information to the next-created run & clears the accumulated information.
+    """
+    def in_context(self):
+        return True
+
+    def tags(self):
+        with _lock:
+            global _table_infos
+            seen = set()
+            unique_infos = []
+            for info in _table_infos:
+                if info not in seen:
+                    unique_infos.append(info)
+                    seen.add(info)
+            try:
```

Comment:
Why do we need try / finally? If we expect failures, should we at least attempt to log tables that do not fail?

Reply:
Good question, we don't need the try-finally, I'll simplify this